x-pack/plugins/lens/public/editor_frame_service/editor_frame/config_panel/config_panel.tsx

```diff
@@ -307,6 +315,22 @@ export function LayerPanels(
                 const datasourcePublicAPI = props.framePublicAPI.datasourceLayers?.[layerId];
                 const datasourceId = datasourcePublicAPI?.datasourceId;
                 dispatchLens(removeDimension({ ...dimensionProps, datasourceId }));
+                if (datasourceId && onUpdateStateCb) {
+                  const nextVisState = activeVisualization.removeDimension({
+                    layerId,
+                    columnId: dimensionProps.columnId,
+                    prevState: visualization.state,
+                    frame: props.framePublicAPI,
+                  });
+                  const activeDatasource = datasourceMap[datasourceId];
+                  const nextDsState = activeDatasource?.removeColumn({
+                    layerId: dimensionProps.layerId,
+                    columnId: dimensionProps.columnId,
+                    prevState: datasourceStates[datasourceId].state,
+                    indexPatterns: props.framePublicAPI.dataViews.indexPatterns,
+                  });
+                  onUpdateStateCb(nextDsState, nextVisState);
```

Comment:
As I understand it, this is a simplified version of the logic in the [`removeDimension` reducer](https://github.com/elastic/kibana/blob/main/x-pack/plugins/lens/public/state_management/lens_slice.ts#L1151). To my mind, it would be nice to keep this aligned with that routine.

Two ideas (not sure if they work or not).

1. could we extract the new state from our store after firing the `removeDimension` action?

```ts
const lensStore = useStore();
...
onRemoveDimension={(dimensionProps) => {
...
    dispatchLens(removeDimension({ ...dimensionProps, datasourceId }));
    if (datasourceId && onUpdateStateCb) {
        const newState = lensStore.getState().lens;
        onUpdateStateCb(selectVisualizationState(newState), selectDatasourceStates(newState)[datasourceId]);
    }
```

2. If that doesn't work, could we export the logic in the `removeDimension` reducer and use it here?

Reply:
Nice call my dearest, done!